File: backend-golang/internal/handlers/admin/uploads.go
package admin

import (
	"encoding/json"
	"net/http"

	"inanhxink/backend-golang/internal/config"
	"inanhxink/backend-golang/internal/handlers"
)

// DELETE /api/admin/uploads — best-effort removal of S3 objects by URL.
func DeleteUploads(w http.ResponseWriter, r *http.Request) {
	var body struct {
		URLs []string `json:"urls"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || len(body.URLs) == 0 {
		handlers.BadRequest(w, "urls (non-empty array) required")
		return
	}

	type result struct {
		URL     string `json:"url"`
		Deleted bool   `json:"deleted"`
		Error   string `json:"error,omitempty"`
	}
	var results []result
	for _, url := range body.URLs {
		deleted, err := config.DeleteFromS3(url)
		r := result{URL: url, Deleted: deleted}
		if err != nil {
			r.Error = err.Error()
		}
		results = append(results, r)
	}
	handlers.OK(w, map[string]any{"success": true, "results": results})
}

File: backend-deprecated/routes/admin/uploads.ts
import { Router, Request, Response } from 'express';
import { sendError } from '../../middleware/sendError';
import { deleteFromS3 } from '../../config/s3';

const router = Router();

// DELETE /api/admin/uploads
// Body: { urls: string[] } — best-effort removal of S3 objects (used to
// clean up orphan files left behind when an admin cancels a pending upload
// or replaces an image without saving).
router.delete('/', async (req: Request, res: Response) => {
  const { urls } = req.body as { urls?: string[] };
  if (!Array.isArray(urls) || urls.length === 0) {
    return res.status(400).json({ success: false, error: 'urls (non-empty array) required' });
  }

  const results: { url: string; deleted: boolean; error?: string }[] = [];
  for (const url of urls) {
    try {
      const deleted = await deleteFromS3(url);
      results.push({ url, deleted });
    } catch (err) {
      results.push({ url, deleted: false, error: (err as Error).message });
    }
  }
  return res.json({ success: true, results });
});

export default router;
